### ISSUE TITLE
Fix handling of output streams, including in cases of error.

### ISSUE THREAD

**Issue by leehambley** [capistrano/sshkit#100]
In SSHKit we do some [weird stuff](https://github.com/capistrano/sshkit/blob/master/lib/sshkit/backends/netssh.rb#L122) with stream handling depending on the packets (sic) we receive back from `Net::SSH`.

I'd like to write a test case script that produces buffered, and unbuffered output, line- and character-wise, on standard error, and standard out to test the correct usage. 

With a pathologically badly behaved script, we could find a way to make this work reliably, I am sure.

**Comment by Aeon** [comment]
So since #86 was closed, what's the currently accepted solution? Seems like there's no way to get stdout/stderr at all right now when a task fails with non-zero exit status?

**Comment by daniel-gomes-sociomantic** [comment]
would also like to know, as redirecting remote command's stdout and stderr to (remote) files and then (remote) `cat`'ing them seems far from ideal

**Comment by robd** [comment]
@daniel-gomes-sociomantic I may have misunderstood the problem here, but there is an option which can be applied to commands to stop non zero commands from raising an exception: ```raise_on_non_zero_exit: false```. The command output is collected in the `full_stderr` and `full_stdout` attrs of the command object.

On SSHKit master, you could patch something like this on to the abstract back end (based on a mixture of the [test](https://github.com/capistrano/sshkit/blob/e06d6b8424979a6a3a2974f375490d2aa15c2ac4/lib/sshkit/backends/abstract.rb#L32) and [capture](https://github.com/capistrano/sshkit/blob/e06d6b8424979a6a3a2974f375490d2aa15c2ac4/lib/sshkit/backends/abstract.rb#L37) methods:

```ruby
module SSHKit
  module Backend
    class Abstract
      def capture_error(*args)
        options = args.extract_options!.merge(raise_on_non_zero_exit: false, verbosity: Logger::DEBUG)
        executed_command = create_command_and_execute(args, options)
        # Do something with executed_command.full_stdout or executed_command.full_stderr
      end
    end
  end
end
```

Does this help?

**Comment by daniel-gomes-sociomantic** [comment]
@robd I'm aware of `raise_on_non_zero_exit: false` (so I might be the one misunderstanding the issue after all), all I wanted was to be able to capture either the `stdout` or the `stderr` of a remote command regardless of whether it succeeded or not (without having to do the redirection myself, that is).

Looking at your proposed patch, it might be just the thing I need!! Unfortunately I am not so ruby-savvy as to being able to make my Capistrano use sshkit's master branch instead of the latest release (v1.7.1)... I tried calling `command` directly from my task but then of course there's no defined `execute_command` function

**Comment by robd** [comment]
@daniel-gomes-sociomantic Are you using a Gemfile with Bundler? If so you can add the following line to your Gemfile:

```ruby
gem 'sshkit', :github => 'capistrano/sshkit', :branch => 'master'
```

and run `bundle install`

**Comment by daniel-gomes-sociomantic** [comment]
> @daniel-gomes-sociomantic Are you using a Gemfile with Bundler?

I'm not, my application is not written in ruby so I never got into the whole Gemfile setup (which is what I meant by not being very Ruby-savvy).

In the meantime, following your idea I manage to get what I wanted by combining the `test` and `capture` methods (from my own codebase) into one:

    module SSHKit
        module Backend
            class Abstract
                def capture_error(*args)
                    options = args.extract_options!.merge(
                        raise_on_non_zero_exit: false,
                        verbosity: Logger::DEBUG
                    )

                    executed_command = _execute(*[*args, options])
    
                    info "stdout: #{executed_command.full_stdout}"
                    info "stderr: #{executed_command.full_stderr}"
                end
            end
        end
    end

and it seems to work fine, cheers @robd!

**Comment by robd** [comment]
@daniel-gomes-sociomantic Ah great - glad you got something working

**Comment by leehambley** [comment]
@mattbrictson we can close this now that Airbrusssh is the default, right ?

**Comment by mattbrictson** [comment]
Yes, this will be fixed at least in Capistrano with the merge of https://github.com/capistrano/capistrano/pull/1541. (I'm waiting until a new version of SSHKit is released to rubygems to move ahead on that PR.)
